File: kmp-algorithm/kmp_algorithm.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<int> computeLPS(const string& pattern) {
    int m = pattern.length();
    vector<int> lps(m, 0);
    int length = 0;
    int i = 1;
    
    while (i < m) {
        if (pattern[i] == pattern[length]) {
            length++;
            lps[i] = length;
            i++;
        } else {
            if (length != 0) {
                length = lps[length - 1];
            } else {
                lps[i] = 0;
                i++;
            }
        }
    }
    return lps;
}

vector<int> kmpSearch(const string& text, const string& pattern) {
    int n = text.length();
    int m = pattern.length();
    
    if (m == 0) return {};
    
    vector<int> lps = computeLPS(pattern);
    vector<int> matches;
    
    int i = 0, j = 0;
    while (i < n) {
        if (pattern[j] == text[i]) {
            i++;
            j++;
        }
        
        if (j == m) {
            matches.push_back(i - j);
            j = lps[j - 1];
        } else if (i < n && pattern[j] != text[i]) {
            if (j != 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }
    
    return matches;
}

int main() {
    string text = "ABABDABACDABABCABCABCABCABC";
    string pattern = "ABABCABCABCABC";
    vector<int> matches = kmpSearch(text, pattern);
    
    cout << "Pattern found at indices: ";
    for (int match : matches) {
        cout << match << " ";
    }
    cout << endl;
    return 0;
}

File: kmp-algorithm/kmp_algorithm.rs
fn compute_lps(pattern: &str) -> Vec<usize> {
    let pattern_bytes = pattern.as_bytes();
    let m = pattern_bytes.len();
    let mut lps = vec![0; m];
    let mut length = 0;
    let mut i = 1;
    
    while i < m {
        if pattern_bytes[i] == pattern_bytes[length] {
            length += 1;
            lps[i] = length;
            i += 1;
        } else if length != 0 {
            length = lps[length - 1];
        } else {
            lps[i] = 0;
            i += 1;
        }
    }
    lps
}

fn kmp_search(text: &str, pattern: &str) -> Vec<usize> {
    let text_bytes = text.as_bytes();
    let pattern_bytes = pattern.as_bytes();
    let n = text_bytes.len();
    let m = pattern_bytes.len();
    
    if m == 0 {
        return Vec::new();
    }
    
    let lps = compute_lps(pattern);
    let mut matches = Vec::new();
    
    let mut i = 0;
    let mut j = 0;
    
    while i < n {
        if pattern_bytes[j] == text_bytes[i] {
            i += 1;
            j += 1;
        }
        
        if j == m {
            matches.push(i - j);
            j = lps[j - 1];
        } else if i < n && pattern_bytes[j] != text_bytes[i] {
            if j != 0 {
                j = lps[j - 1];
            } else {
                i += 1;
            }
        }
    }
    
    matches
}

fn main() {
    let text = "ABABDABACDABABCABCABCABCABC";
    let pattern = "ABABCABCABCABC";
    let matches = kmp_search(text, pattern);
    println!("Pattern found at indices: {:?}", matches);
}
